lib/auth/password.go

```diff
@@ -453,55 +453,17 @@ func (s *Server) changeUserSecondFactor(req *proto.ChangeUserAuthenticationReque
 	if secondFactor == constants.SecondFactorOff {
 		return nil
 	}
-	switch req.NewMFARegisterResponse.GetResponse().(type) {
-	case *proto.MFARegisterResponse_TOTP:
-		if secondFactor == constants.SecondFactorU2F {
-			return trace.BadParameter("user %q sent an OTP token during password reset but cluster only allows U2F for second factor", username)
-		}
-		secrets, err := s.Identity.GetUserTokenSecrets(ctx, req.TokenID)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-
-		dev, err := services.NewTOTPDevice("otp", secrets.GetOTPKey(), s.clock.Now())
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.checkTOTP(ctx, username, req.GetNewMFARegisterResponse().GetTOTP().GetCode(), dev); err != nil {
-			return trace.Wrap(err)
-		}
-		if err := s.UpsertMFADevice(ctx, username, dev); err != nil {
-			return trace.Wrap(err)
-		}
 
-		return nil
-
-	case *proto.MFARegisterResponse_U2F:
-		if secondFactor == constants.SecondFactorOTP {
-			return trace.BadParameter("user %q sent a U2F registration during password reset but cluster only allows OTP for second factor", username)
-		}
-		cfg, err := cap.GetU2F()
-		if err != nil {
-			return trace.Wrap(err)
-		}
+	if req.GetNewMFARegisterResponse() == nil && secondFactor != constants.SecondFactorOptional {
+		return trace.BadParameter("no second factor sent during user %q password reset", username)
+	}
 
-		_, err = u2f.RegisterVerify(ctx, u2f.RegisterVerifyParams{
-			DevName:                "u2f",
-			ChallengeStorageKey:    req.TokenID,
-			RegistrationStorageKey: username,
-			Resp: u2f.RegisterChallengeResponse{
-				RegistrationData: req.GetNewMFARegisterResponse().GetU2F().GetRegistrationData(),
-				ClientData:       req.GetNewMFARegisterResponse().GetU2F().GetClientData(),
-			},
-			Storage:        s.Identity,
-			Clock:          s.GetClock(),
-			AttestationCAs: cfg.DeviceAttestationCAs,
-		})
-		return trace.Wrap(err)
-	default:
-		if secondFactor != constants.SecondFactorOptional {
-			return trace.BadParameter("no second factor sent during user %q password reset", username)
-		}
+	// Default device name still used as UI invite/reset
+	// forms does not allow user to enter own device names yet.
+	deviceName := "u2f"
```

Comment:
Won't this break if they already did this once? Should we go for a randomized name, or at least pick something that we know isn't taken by the user's existing devices?

Reply:
it's safe b/c the `ChangeUserAuthenticationRequest` is a request from a user invite or reset, where the user does not have any mfa devices

Reply:
Add to the comment that we expect the user to have no devices here?